### PR TITLE
Allow mixed declaration and assignment in deconstruction

### PR DESCRIPTION
**Pull request by YairHalberstadt**
Implements https://github.com/dotnet/csharplang/issues/125
Test plan: https://github.com/dotnet/roslyn/issues/47746

I simply removed the error, and added tests. So far everything seems to work without any actual code changes, which worries me a little :-)

Please suggest scenarios for further testing.

### CODE REVIEW

**Comment by RikkiGibson** [comment]
are there any tests for nested deconstruction? i.e. `(var x, (var y, z)) = new C();`

**Comment by RikkiGibson** [comment]
do we need to retarget this to a "compiler-next" branch? @jaredpar

**Comment by YairHalberstadt** [comment]
> are there any tests for nested deconstruction? i.e. `(var x, (var y, z)) = new C();`

Thanks, added a test

**Comment by jcouv** [comment]
> So far everything seems to work without any actual code changes [just removing the error]

That's what we expected :-)

Just as a heads up, we likely won't have much time to review this PR before 6/10 (we're trying to get a big chunk of records in next preview).
Also, we'll want to double-check this feature by LDM. I don't feel strongly opposed, but I still don't much like the idea of mixing from a readability point of view :-/

**Comment by alrz** [comment]
Is there any consideration to permit tuple element names in deconstruction? The expected behavior is already established by recursive patterns for both Deconstruct and tuples. See https://github.com/dotnet/csharplang/issues/3304

**Comment by YairHalberstadt** [comment]
@alrz

I don't see why that's related? This just removes a (somewhat arbitrary) error.

**Comment by alrz** [comment]
>I don't see why that's related? This just removes a (somewhat arbitrary) error.

I believe this is also a syntactical relaxation for deconstruction and removes an error. The parser already accept such code.

**Comment by jcouv** [comment]
LDM approved yesterday (9/15/2020).

File: src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:46]: `Deconstructor` -> `Deconstruction`. I don't think we use the term `deconstructor` anywhere.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:9227]: We probably don't need to verify IL here. It's long and not sure it helps.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 5)

**Comment by RikkiGibson** [comment]
The feature is approved as preview feature for C# 10 (VS 16.9 under LangVersion.Preview). If this PR is to be merged in the next couple of weeks, then it will be necessary to update the base branch to `release/dev16.9-preview1` (tenatively named).

If on the other hand it will be a while before the PR is merged, we can just wait until master represents VS 16.9.

**Comment by YairHalberstadt** [comment]
> The feature is approved as preview feature for C# 10 (VS 16.9 under LangVersion.Preview). If this PR is to be merged in the next couple of weeks, then it will be necessary to update the base branch to release/dev16.9-preview1 (tenatively named).

Would it not be possible to merge it now as a preview feature for C# 10?  All the current C# 9 features are no longer in preview. There's zero risk of breaking C# 9.0 since all this does is replace a warning with a preview gate - no code gen changes at all.

Otherwise might as well wait a couple of weeks.

Either way I imagine that there's further tests that need to be done before this can be merged:

Off the top of my head I'm guessing Symbols and Control flow tests? Anything else? GoToDefinition tests?

**Comment by jcouv** [comment]
Yes, we discussed with Rikki offline and I think we can use the `master` branch right away since the feature is behind LangVer=preview. 

> Either way I imagine that there's further tests that need to be done before this can be merged

Yes. I opened a mini [test-plan issue](https://github.com/dotnet/roslyn/pull/47745) for tracking. I think we can just do the whole feature in a single PR.

For IDE side, I'll start a thread and get back to you.

From the compiler side:
- A data flow test would be good (a test in the vein of `RegionAnalysisTests.AnalysisOfExpressionInTupleEquality`).
- Let's check the semantic model in nested scenario (see `VerifyModelForDeconstructionLocal` which covers declaration, and add checks for `GetTypeInfo` and `GetSymbolInfo` on variables used in mixed declarations).
- Add an IOperation test

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:6141]: Please keep a C# 9 check as well

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:6651]: Why change the LangVer?

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:6671]: Please keep a C# 9 check as well and remove comment below

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:7086]: Please keep a C# 9 check as well

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:9712]: Could you add a C# 9 diagnostics check for nested case as well? Thanks

File: src/Compilers/CSharp/Test/Syntax/Parsing/DeconstructionTests.cs
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/DeconstructionTests.cs:2102]: Let's keep the C# 9 diagnostic check here too

**Comment by jcouv** [comment]
Correction from earlier: Jared recommends to hold off merging until we have a branch for 16.9, to minimize risk. We can leave the PR targeting `master` for now. I'll start feature review process internally (to brainstorm more tests for test plan).

**Comment by jaredpar** [comment]
The 16.9 branch already exists: `/releases/16.9-preview1` so we could just change the merge target now if we want. 

https://github.com/dotnet/roslyn/tree/release/dev16.9-preview1

**Comment by jcouv** [comment]
Done. Retargeted PR to `release/dev16.9-preview1`. Thanks

**Comment by YairHalberstadt** [comment]
> Yes. I opened a mini test-plan issue for tracking. I think we can just do the whole feature in a single PR.

Is the link correct there?

**Comment by jcouv** [comment]
The link was incorrect indeed. Here's the [test plan](https://github.com/dotnet/roslyn/issues/47746) link.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:6202]: Let's VerifyIL for the tests that actually are expected to run.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:7130]: Let's run this test, and verify the IL

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:7363]: Let's run this test, and verify the IL

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:9557]: Please add a similar test for use-before-declaration. ie:
```cs
(x, var y) = new C();
int x;
```

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:397]: I'm glad that our somewhat overengineered design ended up being useful :D

**Comment by 333fred** [comment]
Done review pass (commit 9). Just a couple of test comments, this might be the smallest feature in terms of needed code ever :)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: Sorry, this was a more general comment for the tests below as well

---
In reply to: [490632506](https://github.com/dotnet/roslyn/pull/44476#discussion_r490632506) [](ancestors = 490632506)

**Comment by 333fred** [comment]
Done review pass (commit 10)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (YairHalberstadt) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs:9944]: Added here: https://github.com/dotnet/roslyn/pull/44476/files#diff-7c37d20e00ca6192dae6ed6ead852ab8R9857

**Comment by jcouv** [comment]
Something is wrong with the PR (too many commits). What happended?

**Comment by YairHalberstadt** [comment]
@jcouv I merged from master

**Comment by YairHalberstadt** [comment]
Instead of the target branch

**Comment by RikkiGibson** [comment]
Sorry for the runaround here @YairHalberstadt.. we have snapped the 16.8 release, master is now 16.9, and the 16.9-preview1 branch is currently defunct

**Comment by YairHalberstadt** [comment]
@RikkiGibson well the number of commits seems correct now, so all is good!

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: @YairHalberstadt I think this is my only comment left outstanding.

Comment (YairHalberstadt) [reply src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenDeconstructTests.cs]: I'm now verifying IL for all tests I added modified, except a couple @jcouv said not to because they're too long.

There is a test directly below which github thinks I modified, but that's just the git diff going wrong.

**Review by 333fred — APPROVED** [review]
LGTM (commit 13). I'm not exactly sure when we'll be able to get this merged, hopefully sometime in the next week or 2. We're heads down on remaining C# 9 fixes for the next 2 weeks.

**Comment by YairHalberstadt** [comment]
@333fred any chance we can get this merged this week?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 14)

**Comment by 333fred** [comment]
Next week.

**Comment by jcouv** [comment]
Sorry for the delay. I'm going to run an IDE test pass today and find an IDE buddy.

**Comment by jcouv** [comment]
Ran this change through a quick IDE test pass on Friday, focusing on completion and QuickInfo (since those seems most likely to be affected). Things look good (behavior is same as for regular deconstruction). Pinged Jinu to get an buddy to get the IDE sign-off.

**Comment by sharwell** [comment]
It looks like there are several test gaps in IDE code.

* `VarKeywordRecommenderTests`, `IntKeywordRecommenderTests`, etc., should be updated:
    * Verify keyword is provided for `(x, $$`
    * Verify the keyword is _not_ provided for `var (x, $$`
* `SymbolCompletionProviderTests` needs to be updated to explicitly test `(var x, $$` (reference provided) and `var (x, $$` (reference not provided). See `TupleTypeAtMemberLevel3` for an example and positioning within the file.

**Comment by YairHalberstadt** [comment]
Looks like `(var _, _) = (0, 0);` is allowed today. I have absolutely no idea why, but it would be a breaking change to remove.

**Comment by YairHalberstadt** [comment]
As is:

```csharp
        (var x, _) = (0, 0);
        Console.WriteLine(x);
 ```

So it looks like mixed declaration and discard has always been allowed.

**Comment by YairHalberstadt** [comment]
> Verify that x is not provided for (var x, $$ (while other symbols are)

Currently the logic to check is a local is in scope simply checks that's it's declared before the current syntax, so this is suggested, as is x in `var x  = $$`.

I don't want to change that logic in the PR as it seems like a relatively complex change, and we do this already, so we're no worse off by merging this.

**Comment by sharwell** [comment]
How does the language define this code after your feature?

```csharp
class T {
  int x;
  void M() {
    (var x, x) = (1, 2);
    Console.WriteLine(x);
    Console.WriteLine(this.x);
  }
}
```

**Comment by YairHalberstadt** [comment]
@sharwell that's an error

**Comment by sharwell** [comment]
And `(x, var x)` there would also be an error?

**Comment by YairHalberstadt** [comment]
Yes, and I have tests for both

**Comment by jcouv** [comment]
@sharwell We have the necessary sign-offs from the compiler side. Yair has added the IDE (thanks!) and they match what I'd verified manually (behave as expected). 
We're ready to merge if you're all good. Thanks

**Review by sharwell — APPROVED** [review]

**Comment by jcouv** [comment]
Merged/squashed. Thanks @YairHalberstadt! :-)

**Pull request merged**
